### PR TITLE
Update: Use glob for ignored files and remove minimatch (fixes #1095)

### LINKED ISSUE #1095: Performance: Pre-compile minimatch patterns
minimatch has a method `makeRe` that will precompile a pattern into a regexp. Using this should greatly increase ignored-paths performance, especially on large numbers of patterns and/or file paths. It would also eliminate any overhead from comment patterns. I have not yet tested any performance difference - will do so sometime soon.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3122339-performance-pre-compile-minimatch-patterns?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gyandeeps**

### CODE REVIEW

**Comment by ilyavolodin** [comment]
Nice. Build is failing on one of the unittests due to full file path being returned. Can you also check performance difference please?

**Comment by gyandeeps** [comment]
That unit test is not failing on my machine (windows 7). And I have not changed any path for `processOnText` function.

**Comment by ilyavolodin** [comment]
Let me try it locally.

**Comment by ilyavolodin** [comment]
It's failing on my Windows 8 machine with git bash.

File: lib/cli-engine.js [hunk 1]
```diff
@@ -536,12 +536,18 @@ CLIEngine.prototype = {
...
+
+        globOptions = {
+            nodir: true,
+            realpath: true,
+            ignore: ignoredPaths.map(processPath)
+        };
 
-        if (filename && options.ignore && exclude(filename)) {
+        if (filename && options.ignore && glob.sync(filename, globOptions).length === 0) {
             results.push(createIgnoreResult(filename));
```
Comment (ilyavolodin) [inline lib/cli-engine.js]: This is where the unittest fails. createIgnoreResults does `path.resolve(filename)`. Not sure why it's only failing now though.

File: lib/cli-engine.js [hunk 2]
```diff
@@ -536,12 +536,18 @@ CLIEngine.prototype = {
...
+            ignoredPaths = options.ignore ? IgnoredPaths.load(options).patterns : [];
+
+        globOptions = {
+            nodir: true,
+            realpath: true,
+            ignore: ignoredPaths.map(processPath)
+        };
 
-        if (filename && options.ignore && exclude(filename)) {
+        if (filename && options.ignore && glob.sync(filename, globOptions).length === 0) {
```
Comment (lo1tuma) [inline lib/cli-engine.js]: Isn’t this more inefficient than the previous `minimatch` implementation? Looks like `glob.sync` is a blocking I/O function and `minimatch` only compares strings via regexp.

File: lib/cli-engine.js (hunk 1 above)
Comment (gyandeeps) [inline lib/cli-engine.js]: on my windows 7, else branch gets executed for the unit test we are talking about.

**Comment by gyandeeps** [comment]
Perf

**Current master:**
``sh
gs025879@WIN1551863 /C/Users/gs025879/Documents/webstrom/eslint (master)
$ npm run perf

> eslint@1.3.1 perf c:\Users\gs025879\Documents\webstrom\eslint
> node Makefile.js perf

CPU Speed is 2793 with multiplier 7500000
Performance Run #1:  2216.467073ms
Performance Run #2:  2275.665615ms
Performance Run #3:  2160.722231ms
Performance Run #4:  2190.225841ms
Performance Run #5:  2204.95987ms
Performance budget ok:  2204.95987ms (limit: 2685.2846401718584ms)
``

**This branch:**
```sh
gs025879@WIN1551863 /C/Users/gs025879/Documents/webstrom/eslint (issue1095)
$ npm run perf

> eslint@1.3.1 perf c:\Users\gs025879\Documents\webstrom\eslint
> node Makefile.js perf

CPU Speed is 2793 with multiplier 7500000
Performance Run #1:  2194.400116ms
Performance Run #2:  2167.220964ms
Performance Run #3:  2161.738953ms
Performance Run #4:  2151.532522ms
Performance Run #5:  2179.763947ms
Performance budget ok:  2167.220964ms (limit: 2685.2846401718584ms)
```

**Comment by lo1tuma** [comment]
The changes made in this PR don't affect the result of `npm run perf` because it lints only a single file without any files being ignored.

**Comment by gyandeeps** [comment]
I think you are correct @lo1tuma that glob is IO bound.

**Comment by lo1tuma** [comment]
BTW: If we are going to use the [`Ignore`](https://www.npmjs.com/package/ignore) module to solve #2365 then #1095 would become obsolete anyway.

**Comment by gyandeeps** [comment]
Oh yeah. I should close this PR then. No point going down this path now.
Should we close #1095 in favor of #2365 ? @eslint/eslint-team

**Comment by ilyavolodin** [comment]
Yes, let's close #1095

**Comment by nzakas** [comment]
So can this be closed as well?

**Comment by gyandeeps** [comment]
Yes. I am going to close both.

**Pull request closed without merging**